tools/PI/DevHome.PI/ViewModels/BarWindowViewModel.cs

```diff
@@ -236,4 +272,255 @@ private void PerfCounterHelper_PropertyChanged(object? sender, PropertyChangedEv
             });
         }
     }
+
+    // External tools
+    private void InitializeExternalTools()
+    {
+        ExternalToolsHelper.Instance.Init();
+
+        foreach (var tool in ExternalToolsHelper.Instance.AllExternalTools)
+        {
+            tool.PropertyChanged += ExternalToolItem_PropertyChanged;
+        }
+
+        // We have to cast to INotifyCollectionChanged explicitly because the CollectionChanged
+        // event in ReadOnlyObservableCollection is protected.
+        ((INotifyCollectionChanged)ExternalToolsHelper.Instance.AllExternalTools).CollectionChanged += ExternalTools_CollectionChanged;
+    }
+
+    private void ExternalTools_CollectionChanged(object? sender, NotifyCollectionChangedEventArgs e)
+    {
+        _dispatcher.TryEnqueue(() =>
+        {
+            if (e.Action == NotifyCollectionChangedAction.Add && e.NewItems is not null)
+            {
+                foreach (ExternalTool tool in e.NewItems)
+                {
+                    foreach (MenuFlyout flyout in _externalToolMenus)
+                    {
+                        AddExternalToolToContextMenu(flyout, tool);
+                    }
+
+                    // Listen for tool changes
+                    tool.PropertyChanged += ExternalToolItem_PropertyChanged;
+                }
+            }
+            else if (e.Action == NotifyCollectionChangedAction.Remove && e.OldItems is not null)
+            {
+                foreach (ExternalTool tool in e.OldItems)
+                {
+                    tool.PropertyChanged -= ExternalToolItem_PropertyChanged;
+
+                    foreach (MenuFlyout flyout in _externalToolMenus)
+                    {
+                        RemoveExternalToolFromContextMenu(flyout, tool);
+                    }
+                }
+            }
+        });
+    }
+
+    private void AddExternalToolToContextMenu(MenuFlyout menuFlyout, ExternalTool tool)
+    {
+        menuFlyout.Items.Add(CreateContextMenuItemForTool(tool));
+    }
+
+    private void RemoveExternalToolFromContextMenu(MenuFlyout menuFlyout, ExternalTool tool)
+    {
+        foreach (var menuItem in menuFlyout.Items)
+        {
+            if (menuItem.Tag == tool)
+            {
+                tool.PropertyChanged -= ExternalToolItem_PropertyChanged;
+                menuFlyout.Items.Remove(menuItem);
+                break;
+            }
+        }
+    }
+
+    // This creates a Menu Flyout item for an external tool. It also creates a sub-menu item for pinning/unpinning
+    // and unregistering the tool.
+    private MenuFlyoutItem CreateContextMenuItemForTool(ExternalTool tool)
+    {
+        var imageIcon = new ImageIcon
+        {
+            Source = tool.ToolIcon,
+        };
```

Comment:
Is it possible to create a MenuFlyoutItemControl and use it here?

Reply:
Can you elaborate?

Reply:
Sure, can work on it once this PR is in.

Reply:
Makes sense, and it did simplify some things as well.